yt_dlp/extractor/youtube.py

```diff
@@ -5765,77 +5897,94 @@ def get_mobj(url):
-                            redirect_warning += f'. {selected_tab_name} tab is being downloaded instead'
+            selected_tab_id, selected_tab_name = self._extract_tab_id_and_name(selected_tab, url)  # NB: Name may be translated
+            requested_tab_id = mobj['tab'][1:]
+            self.write_debug(f'Selected tab: {selected_tab_id!r} ({selected_tab_name}), '
+                             f'Requested tab: {requested_tab_id!r}, Original tab: {original_tab_id!r}')
+
+            if not original_tab_id and selected_tab_name:
+                if self._has_tab(tabs, 'streams'):
+                    tab_results.append(self.url_result(''.join((pre, '/streams', post))))
+                if self._has_tab(tabs, 'shorts'):
+                    tab_results.append(self.url_result(''.join((pre, '/shorts', post))))
+                # XXX: Members-only tab should also be extracted
+
+                if not tab_results and selected_tab_id != 'videos':
+                    # Channel does not have streams, shorts or videos tabs
+                    if item_id[:2] != 'UC':
+                        raise ExtractorError('This channel has no uploads', expected=True)
+
+                    # Topic channels don't have /videos. Use the equivalent playlist instead
+                    pl_id = f'UU{item_id[2:]}'
+                    pl_url = f'https://www.youtube.com/playlist?list={pl_id}'
+                    try:
+                        data, ytcfg = self._extract_data(pl_url, pl_id, ytcfg=ytcfg, fatal=True, webpage_fatal=True)
+                    except ExtractorError:
+                        raise ExtractorError('This channel has no uploads', expected=True)
                     else:
-                        raise ExtractorError(redirect_warning, expected=True)
+                        item_id, url, selected_tab_name = pl_id, pl_url, requested_tab_id
+                        self.to_screen(
+                            f'The channel does not have a videos, shorts, or live tab. Redirecting to playlist {pl_id} instead')
+
+                elif tab_results and selected_tab_id != 'videos':
+                    # When there are shorts/live tabs but not videos tab
+                    url, data = ''.join((pre, post)), None
+
+            elif requested_tab_id != selected_tab_id:
+                if requested_tab_id == 'live':
+                    # Live tab should have redirected to the video
+                    # Except in the case the channel has an actual live tab
+                    # Example: https://www.youtube.com/channel/UCEH7P7kyJIkS_gJf93VYbmg/live
+                    raise UserNotLive(video_id=mobj['id'])
+                elif selected_tab_name:
+                    raise ExtractorError(f'This channel does not have a {requested_tab_id} tab', expected=True)
+
+                # For channels such as https://www.youtube.com/channel/UCtFRv9O2AHqOZjjynzrv-xg
+                url = ''.join((pre, f'/{selected_tab_id}' if selected_tab_id else '', post))
```

Comment:
pls explain

Reply:
this is for the case where we accept the home/default page. The tab url here would be /videos but we are downloading a different tab so we need to fix it.

Though the behaviour here rewrites it as https://www.youtube.com/channel/UCtFRv9O2AHqOZjjynzrv-xg/UCtFRv9O2AHqOZjjynzrv-xg. Whether that is a valid tab url on youtube's side is unknown, but yt-dlp will treat that as a direct tab url.

Reply:
why not just strip the `/videos`?
```
url = f'{pre}{post}'
```

Reply:
was just trying to be general. that works too, but by yt-dlp that isn't really the correct url